dbcon/mysql/ha_mcs_execplan.cpp

```diff
@@ -6544,25 +6544,176 @@ int processWhere(SELECT_LEX &select_lex,
+
+            if (select_lex.select_limit)
+                csep->limitNum(((Item_int*)select_lex.select_limit)->val_int());
+        }
+    }
+    // other types of queries that have explicit LIMIT
+    else if (select_lex.explicit_limit)
+    {
+        uint32_t limitOffset = 0;
+
+        if (select_lex.join)
+        {
+            JOIN* join = select_lex.join;
+#if MYSQL_VERSION_ID >= 50172
+
+            // @bug5729. After upgrade, join->unit sometimes is uninitialized pointer
+            // (not null though) and will cause seg fault. Prefer checking
+            // select_lex->offset_limit if not null.
+            if (join->select_lex &&
+                    join->select_lex->offset_limit &&
+                    join->select_lex->offset_limit->is_fixed() &&
+                    join->select_lex->select_limit &&
+                    join->select_lex->select_limit->is_fixed())
+            {
+                limitOffset = join->select_lex->offset_limit->val_int();
+                limitNum = join->select_lex->select_limit->val_int();
+            }
+            else if (join->unit)
+            {
+                limitOffset = join->unit->offset_limit_cnt;
+                limitNum = join->unit->select_limit_cnt - limitOffset;
+            }
+
+#else
+            limitOffset = (join->unit)->offset_limit_cnt;
+            limitNum = (join->unit)->select_limit_cnt - (join->unit)->offset_limit_cnt;
+#endif
+        }
+        else
+        {
+            if (select_lex.master_unit()->global_parameters()->offset_limit)
+            {
+                Item_int* offset = (Item_int*)select_lex.master_unit()->global_parameters()->offset_limit;
+                limitOffset = offset->val_int();
+            }
+
+            if (select_lex.master_unit()->global_parameters()->select_limit)
+            {
+                Item_int* select = (Item_int*)select_lex.master_unit()->global_parameters()->select_limit;
+                limitNum = select->val_int();
+            }
+        }
+
+        csep->limitStart(limitOffset);
+        csep->limitNum(limitNum);
+    }
+    // If an explicit limit is not specified, use the system variable value
+    else
+    {
+        csep->limitNum(gwi.thd->variables.select_limit);
+    }
+
+    // We don't currently support limit with correlated subquery
+    if (gwi.subQuery && !gwi.correlatedTbNameVec.empty() && csep->hasOrderBy())
+    {
+        gwi.fatalParseError = true;
+        gwi.parseErrorText = IDBErrorInfo::instance()->errorMsg(ERR_NON_SUPPORT_LIMIT_SUB);
+        setError(gwi.thd, ER_INTERNAL_ERROR, gwi.parseErrorText, gwi);
+        return ER_CHECK_NOT_IMPLEMENTED;
+    }
+
+    // MDB applies OFFSET on its own for SH processing.
+    // See related MDEV-16327 for details.
+    if (isSelectHandlerTop && csep->limitStart())
+    {
+        if (std::numeric_limits<uint64_t>::max()-csep->limitStart()
+            < csep->limitStart())
```

Comment:
I think the intention here is to check the overflow for the sum of csep->limitStart() and csep->limitNum().